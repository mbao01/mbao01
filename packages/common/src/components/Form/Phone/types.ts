import { UsePhoneInputConfig } from "react-international-phone";
import { type InputProps } from "../Input/types";

export type PhoneProps = Omit<InputProps, "type" | "label" | "inputMode"> &
  Pick<UsePhoneInputConfig, "defaultCountry">;
