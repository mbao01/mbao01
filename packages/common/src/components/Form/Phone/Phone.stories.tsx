import type { Meta, StoryObj } from "@storybook/react-vite";
import { Phone } from "./Phone";

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  component: Phone,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: "centered",
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ["autodocs"],
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {},
} satisfies Meta<typeof Phone>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Default: Story = {
  args: {
    "aria-label": "Phone number",
  },
};

export const DefaultCountryPhone: Story = {
  args: {
    "aria-label": "Phone number",
    defaultCountry: "ng",
  },
};

export const PrimaryPhone: Story = {
  args: {
    variant: "primary",
    "aria-label": "Phone number",
  },
};

export const OutlinePhone: Story = {
  args: {
    outline: true,
    variant: "accent",
    "aria-label": "Phone number",
  },
};

export const TinyPhone: Story = {
  args: {
    size: "xs",
    variant: "warning",
    placeholder: "Enter phone number",
  },
};

export const DisabledPhone: Story = {
  args: {
    disabled: true,
    "aria-label": "Phone number",
  },
};
