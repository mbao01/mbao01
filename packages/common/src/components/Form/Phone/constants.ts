import { cva } from "../../../libs";

export const getPhoneButtonClasses = cva("join-item justify-between px-2");

export const getPhoneInputClasses = cva(
  "input join-item input-bordered w-full !rounded-l-none !pl-2 transition-all duration-100"
);
