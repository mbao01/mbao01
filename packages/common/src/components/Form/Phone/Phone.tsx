"use client";

import React, { MutableRefObject } from "react";
import {
  defaultCountries,
  usePhoneInput,
  FlagImage,
} from "react-international-phone";
import { type PhoneProps } from "./types";
import { Popover } from "../../Popover";
import { Command } from "../../Command";
import { cn } from "../../../utilities";
import { Button } from "../../Button";
import { Input } from "../Input";
import { getPhoneButtonClasses, getPhoneInputClasses } from "./constants";

const Phone = React.forwardRef<HTMLInputElement, PhoneProps>(
  (
    {
      disabled,
      defaultValue,
      className,
      placeholder,
      onChange,
      variant,
      outline,
      wide,
      size,
      ...props
    },
    ref
  ) => {
    const [open, setOpen] = React.useState(false);
    const {
      inputValue,
      handlePhoneValueChange,
      inputRef,
      country,
      setCountry,
    } = usePhoneInput({
      value: String(defaultValue) ?? "+231",
      defaultCountry: "lr",
      forceDialCode: true,
      inputRef: ref as MutableRefObject<HTMLInputElement | null>,
    });

    const options = defaultCountries.map((country) => ({
      label: country[0],
      value: country[1],
    }));

    const handleInputChange = (e: React.ChangeEvent<HTMLInputElement>) => {
      handlePhoneValueChange(e);
      onChange?.(e);
    };

    return (
      <div className="join">
        <Popover open={open} onOpenChange={setOpen}>
          <Popover.Trigger asChild>
            <Button
              role="combobox"
              aria-expanded={open}
              size={size}
              variant={variant}
              outline={outline}
              className={cn(getPhoneButtonClasses())}
              disabled={disabled}
            >
              <FlagImage
                iso2={country.iso2}
                size={["xs", "sm"].includes(size!) ? "20px" : "30px"}
              />
            </Button>
          </Popover.Trigger>
          <Popover.Content className="w-[200px] p-0">
            <Command>
              <Command.Input placeholder="Search country..." />
              <Command.Empty>No country found.</Command.Empty>
              <Command.List>
                <Command.Group>
                  {options.map((option) => (
                    <Command.Item
                      key={option.value}
                      value={option.value}
                      keywords={[option.label, option.value]}
                      onSelect={(currentValue) => {
                        setCountry(currentValue);
                        setOpen(false);
                      }}
                      className={cn({
                        "bg-base-300": country.iso2 === option.value,
                      })}
                    >
                      <FlagImage iso2={option.value} size="20px" />
                      <span className="truncate ml-2">{option.label}</span>
                    </Command.Item>
                  ))}
                </Command.Group>
              </Command.List>
            </Command>
          </Popover.Content>
        </Popover>
        <Input
          type="text"
          ref={inputRef}
          value={inputValue}
          pattern="^\+\d{1,4}\s\d{6,}$"
          inputMode="numeric"
          placeholder={placeholder ?? " "}
          onChange={handleInputChange}
          size={size}
          wide={wide}
          variant={variant}
          outline={outline}
          disabled={disabled}
          className={cn(getPhoneInputClasses(), className)}
          {...props}
        />
      </div>
    );
  }
);
Phone.displayName = "Phone";

export { Phone };
