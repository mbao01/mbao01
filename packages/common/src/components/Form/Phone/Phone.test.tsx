import { describe, expect, it } from "vitest";
import { render, screen } from "@testing-library/react";
import { Phone } from "./Phone";
import userEvent from "@testing-library/user-event";

describe("Phone", () => {
  it("should render a phone input", () => {
    const { asFragment } = render(<Phone name="mobile" />);

    expect(screen.getByRole("textbox")).toHaveValue("+231 ");
    expect(screen.getByRole("combobox")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("should input phone number", async () => {
    const user = userEvent.setup();
    const { asFragment } = render(
      <Phone name="mobile" label="Mobile number" />
    );

    const phoneInput = screen.getByRole("textbox");

    await user.type(phoneInput, "9876543210");

    expect(phoneInput).toHaveDisplayValue("+231 9876543210");
    expect(asFragment()).toMatchSnapshot();
  });
});
