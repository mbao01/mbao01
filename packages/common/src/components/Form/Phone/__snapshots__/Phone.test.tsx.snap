// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Phone > should input phone number 1`] = `
<DocumentFragment>
  <div
    class="join"
  >
    <button
      aria-controls="radix-«r1»"
      aria-expanded="false"
      aria-haspopup="dialog"
      aria-label="Select country code"
      class="btn min-h-fit h-10 rounded-md join-item justify-between rounded-r-none! px-2"
      data-state="closed"
      role="combobox"
      type="button"
    >
      <img
        alt=""
        class="react-international-phone-flag-emoji"
        data-country="lr"
        draggable="false"
        height="30px"
        loading="lazy"
        src="https://cdnjs.cloudflare.com/ajax/libs/twemoji/14.0.2/svg/1f1f1-1f1f7.svg"
        style="width: 30px; height: 30px;"
        width="30px"
      />
    </button>
    <input
      class="input rounded-md min-h-fit h-10 input join-item input-bordered w-full rounded-l-none! pl-2! transition-all duration-100"
      inputmode="numeric"
      name="mobile"
      pattern="^\\+\\d{1,4}\\s\\d{6,}$"
      placeholder=" "
      type="text"
      value="+231 9876543210"
    />
  </div>
</DocumentFragment>
`;

exports[`Phone > should render a phone input 1`] = `
<DocumentFragment>
  <div
    class="join"
  >
    <button
      aria-controls="radix-«r0»"
      aria-expanded="false"
      aria-haspopup="dialog"
      aria-label="Select country code"
      class="btn min-h-fit h-10 rounded-md join-item justify-between rounded-r-none! px-2"
      data-state="closed"
      role="combobox"
      type="button"
    >
      <img
        alt=""
        class="react-international-phone-flag-emoji"
        data-country="lr"
        draggable="false"
        height="30px"
        loading="lazy"
        src="https://cdnjs.cloudflare.com/ajax/libs/twemoji/14.0.2/svg/1f1f1-1f1f7.svg"
        style="width: 30px; height: 30px;"
        width="30px"
      />
    </button>
    <input
      class="input rounded-md min-h-fit h-10 input join-item input-bordered w-full rounded-l-none! pl-2! transition-all duration-100"
      inputmode="numeric"
      name="mobile"
      pattern="^\\+\\d{1,4}\\s\\d{6,}$"
      placeholder=" "
      type="text"
      value="+231 "
    />
  </div>
</DocumentFragment>
`;

exports[`Phone > should select country code 1`] = `
<DocumentFragment>
  <div
    class="join"
  >
    <button
      aria-controls="radix-«rdq»"
      aria-expanded="false"
      aria-haspopup="dialog"
      aria-label="Select country code"
      class="btn min-h-fit h-10 rounded-md join-item justify-between rounded-r-none! px-2"
      data-state="closed"
      role="combobox"
      type="button"
    >
      <img
        alt=""
        class="react-international-phone-flag-emoji"
        data-country="af"
        draggable="false"
        height="30px"
        loading="lazy"
        src="https://cdnjs.cloudflare.com/ajax/libs/twemoji/14.0.2/svg/1f1e6-1f1eb.svg"
        style="width: 30px; height: 30px;"
        width="30px"
      />
    </button>
    <input
      class="input rounded-md min-h-fit h-10 input join-item input-bordered w-full rounded-l-none! pl-2! transition-all duration-100"
      inputmode="numeric"
      name="mobile"
      pattern="^\\+\\d{1,4}\\s\\d{6,}$"
      placeholder=" "
      type="text"
      value="+93 "
    />
  </div>
</DocumentFragment>
`;

exports[`Phone > should select country code by searching country abbreviation 1`] = `
<DocumentFragment>
  <div
    class="join"
  >
    <button
      aria-controls="radix-«r6u»"
      aria-expanded="false"
      aria-haspopup="dialog"
      aria-label="Select country code"
      class="btn min-h-fit h-10 rounded-md join-item justify-between rounded-r-none! px-2"
      data-state="closed"
      role="combobox"
      type="button"
    >
      <img
        alt=""
        class="react-international-phone-flag-emoji"
        data-country="gb"
        draggable="false"
        height="30px"
        loading="lazy"
        src="https://cdnjs.cloudflare.com/ajax/libs/twemoji/14.0.2/svg/1f1ec-1f1e7.svg"
        style="width: 30px; height: 30px;"
        width="30px"
      />
    </button>
    <input
      class="input rounded-md min-h-fit h-10 input join-item input-bordered w-full rounded-l-none! pl-2! transition-all duration-100"
      inputmode="numeric"
      name="mobile"
      pattern="^\\+\\d{1,4}\\s\\d{6,}$"
      placeholder=" "
      type="text"
      value="+44 9876 543210"
    />
  </div>
</DocumentFragment>
`;

exports[`Phone > should select country code by searching country name 1`] = `
<DocumentFragment>
  <div
    class="join"
  >
    <button
      aria-controls="radix-«r2»"
      aria-expanded="false"
      aria-haspopup="dialog"
      aria-label="Select country code"
      class="btn min-h-fit h-10 rounded-md join-item justify-between rounded-r-none! px-2"
      data-state="closed"
      role="combobox"
      type="button"
    >
      <img
        alt=""
        class="react-international-phone-flag-emoji"
        data-country="ng"
        draggable="false"
        height="30px"
        loading="lazy"
        src="https://cdnjs.cloudflare.com/ajax/libs/twemoji/14.0.2/svg/1f1f3-1f1ec.svg"
        style="width: 30px; height: 30px;"
        width="30px"
      />
    </button>
    <input
      class="input rounded-md min-h-fit h-10 input join-item input-bordered w-full rounded-l-none! pl-2! transition-all duration-100"
      inputmode="numeric"
      name="mobile"
      pattern="^\\+\\d{1,4}\\s\\d{6,}$"
      placeholder=" "
      type="text"
      value="+234 9876543210"
    />
  </div>
</DocumentFragment>
`;
